### PR TITLE
Virtual entity

### PR DESCRIPTION
**Pull request by SteRiccio**

### CODE REVIEW

File: webapp/loggedin/surveyViews/nodeDefsSelector/components/entitySelector.js [hunk 1]
```diff
@@ -8,14 +8,14 @@ import { nbsp } from '@core/stringUtils'
...
-const getEntities = (hierarchy, lang) => {
+const _generateDropdownItems = (hierarchy, lang) => {
   const entities = []
 
   const traverse = (nodeDef, depth) => {
     const label = NodeDef.getLabel(nodeDef, lang)
     entities.push({
       key: NodeDef.getUuid(nodeDef),
-      value: nbsp + R.repeat(nbsp + nbsp, depth).join('') + label,
+      value: `${nbsp}${R.repeat(nbsp + nbsp, depth).join('')}${label}${NodeDef.isAnalysis(nodeDef) ? ' (V)' : ''}`,
```
Comment (minotogna) [inline webapp/loggedin/surveyViews/nodeDefsSelector/components/entitySelector.js]: I have seen adding the (V) also in Survey.getHierarchy. can we add this instead only once to NodeDef.getLabel?

File: core/survey/_survey/surveyNodeDefs.js [hunk 2]
```diff
@@ -21,15 +21,26 @@ export const getNodeDefsByUuids = (uuids = []) =>
...
+const _getEntitySource = nodeDef => survey =>
+  R.pipe(NodeDef.getEntitySourceUuid, sourceUuid => getNodeDefByUuid(sourceUuid)(survey))(nodeDef)
+
+export const getNodeDefChildren = (nodeDef, includeAnalysis = false) => survey =>
+  R.ifElse(
+    // If nodeDef is virtual, get children from its source
+    R.always(NodeDef.isVirtual(nodeDef)),
+    R.pipe(
+      _getEntitySource(nodeDef),
+      R.ifElse(R.isNil, R.always([]), entitySource => getNodeDefChildren(entitySource, includeAnalysis)(survey)),
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply core/survey/_survey/surveyNodeDefs.js]: solved

File: server/modules/nodeDef/repository/nodeDefRepository.js [hunk 3]
```diff
@@ -82,7 +83,7 @@ export const fetchRootNodeDef = async (surveyId, draft, client = db) =>
   await client.one(
     `SELECT ${nodeDefSelectFields}
      FROM ${getSurveyDBSchema(surveyId)}.node_def 
-     WHERE parent_uuid IS NULL`,
+     WHERE parent_uuid IS NULL AND NOT virtual`,
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply server/modules/nodeDef/repository/nodeDefRepository.js]: right

File: webapp/loggedin/modules/analysis/processingStep/actions.js [hunk 4]
```diff
@@ -101,28 +100,27 @@ export const deleteProcessingStep = () => async (dispatch, getState) => {
...
+    true,
+    true,
+  )
 
   await dispatch({ type: nodeDefCreate, nodeDef })
 
   history.push(`${appModuleUri(analysisModules.nodeDef)}${NodeDef.getUuid(nodeDef)}/`)
 }
+
+export const editEntityVirtual = history => async (_, getState) => {
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/modules/analysis/processingStep/actions.js]: ok, I'll create a common action navigateToNodeDefEdit (history, nodeDefUui) in analysis/actions

File: webapp/loggedin/modules/analysis/processingStep/processingStepView.js [hunk 5]
```diff
@@ -84,14 +90,26 @@ const ProcessingStepView = props => {
...
+                    aria-disabled={hasCalculationSteps}
+                  >
+                    <span className="icon icon-plus icon-12px icon-left" />
+                    {i18n.t('processingStepView.virtualEntity')}
+                  </button>
+                  <button
+                    className="btn btn-s btn-edit"
+                    onClick={() => editEntityVirtual(history)}
+                    aria-disabled={
+                      hasCalculationSteps || (Boolean(processingStepEntity) && !NodeDef.isVirtual(processingStepEntity))
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/modules/analysis/processingStep/processingStepView.js]: right, the condition should be: 
hasCalculationSteps || !processingStepEntity || !NodeDef.isVirtual(processingStepEntity) 
that means that the button is disabled if:
- there are calculation steps
- no entity has been selected
- the selected entity is not a virtual entity

File: webapp/loggedin/modules/analysis/processingStepCalculation/reducer.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ const actionHandlers = {
   // Update calculation attribute on node def edit cancel / back
   [nodeDefSave]: (state, { nodeDef }) =>
     R.when(
-      R.always(NodeDef.isAnalysis(nodeDef)),
+      R.always(NodeDef.isAnalysis(nodeDef) && !NodeDef.isVirtual(nodeDef)),
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/modules/analysis/processingStepCalculation/reducer.js]: ok

File: webapp/loggedin/surveyViews/nodeDefsSelector/components/entitySelector.js (hunk 1 above)
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDefsSelector/components/entitySelector.js]: ok

**Pull request merged**